Allow wlk files in different folders

tesonep opened the pull request:
Fixes #946

fdodino commented:
Wow! Amazing! I am currently working in another branch. As soon as I finish I'll do the review.

fdodino commented:
Hm... I made this project:
https://github.com/wollok/pruebaMultiplesDirectorios

Please, checkout that

* if I use a class defined in a package (Arma), a File not found appears

```
Caused by: java.lang.NullPointerException
	at java.io.File.<init>(File.java:277)
	at org.uqbar.project.wollok.scoping.WollokGlobalScopeProvider.generateUri(WollokGlobalScopeProvider.java:169)
	at org.uqbar.project.wollok.scoping.WollokGlobalScopeProvider.toResource(WollokGlobalScopeProvider.java:135)
	... 55 more
```

That lines are commented, just to avoid receiving such errors in editor.

* If I use a class without defining a package (Guerrero), a runtime error happens (but you get no compile errors)

I think Package = Folder.

fdodino commented:
And by the way, what about #186 and #346, I think there are related...

tesonep commented:
I have added the ability to have relative imports. Maybe it is a hack, but I have added the lookup of all the possible imports.
@fdodino could you check if that fits you? I have added also a test reproducing the case you were commenting before.

fdodino commented:
Grosso!!

I'll finish first with validator issues (give me 1 or 2 days) and I'll test your solution.
Many many many Thanks!!!!!

tesonep commented:
I will continue testing it, I have added the relative imports, and now is not testing all the possible combinations.

fdodino commented:
Hi Pablo,
     I'm sorry but I can't figure out how to make it work. I ran Wollok Environment in Development Mode, and as I had some problems, I decided to delete the whole project and start a new fresh one. This is what I have

![image](https://user-images.githubusercontent.com/4549002/27843600-cd08a450-60eb-11e7-92c3-5150747a5462.png)


I'm not sure what I'm doing wrong, but if I write "import " and then press Ctrl + Space I can see no packages (I added some packages from Wollok environment, but I think it is the same if you create a Folder in the IDE).
If you want me to do further tests, I can try another situation. 

If I try to do

```xtend
import ../aves.aves.*
```

I get a 

```
org.eclipse.emf.common.util.WrappedException: java.lang.ArrayIndexOutOfBoundsException
	at org.eclipse.xtext.linking.lazy.LazyLinkingResource.getEObject(LazyLinkingResource.java:230)
	at org.eclipse.xtext.linking.lazy.LazyLinkingResource.doResolveLazyCrossReference(LazyLinkingResource.java:189)
	at org.eclipse.xtext.linking.lazy.LazyLinkingResource.resolveLazyCrossReference(LazyLinkingResource.java:148)
	at org.eclipse.xtext.linking.lazy.LazyLinkingResource.resolveLazyCrossReferences(LazyLinkingResource.java:134)
	at org.eclipse.xtext.EcoreUtil2.resolveLazyCrossReferences(EcoreUtil2.java:498)
	at org.eclipse.xtext.resource.impl.DefaultResourceDescription.computeReferenceDescriptions(DefaultResourceDescription.java:135)
	at org.eclipse.xtext.resource.impl.DefaultResourceDescription$2.get(DefaultResourceDescription.java:123)
	at org.eclipse.xtext.resource.impl.DefaultResourceDescription$2.get(DefaultResourceDescription.java:120)
	at org.eclipse.xtext.util.OnChangeEvictingCache.get(OnChangeEvictingCache.java:77)
	at org.eclipse.xtext.resource.impl.DefaultResourceDescription.getReferenceDescriptions(DefaultResourceDescription.java:120)
	at org.eclipse.xtext.resource.DescriptionUtils.collectOutgoingReferences(DescriptionUtils.java:29)
	at org.eclipse.xtext.resource.impl.DefaultResourceDescriptionManager.isAffected(DefaultResourceDescriptionManager.java:117)
	at org.eclipse.xtext.ui.editor.DirtyStateEditorSupport.isReparseRequired(DirtyStateEditorSupport.java:649)
	at org.eclipse.xtext.ui.editor.DirtyStateEditorSupport$UpdateEditorStateJob$1.exec(DirtyStateEditorSupport.java:161)
	at org.eclipse.xtext.ui.editor.DirtyStateEditorSupport$UpdateEditorStateJob$1.exec(DirtyStateEditorSupport.java:1)
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:91)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:520)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.readOnly(XtextDocument.java:492)
	at org.eclipse.xtext.ui.editor.model.XtextDocument.readOnly(XtextDocument.java:133)
	at org.eclipse.xtext.ui.editor.DirtyStateEditorSupport$UpdateEditorStateJob.run(DirtyStateEditorSupport.java:146)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
Caused by: java.lang.ArrayIndexOutOfBoundsException
```

tesonep commented:
Fuck, I will continue trying it.

Thanks Fer!!!

Cheers.

On Wed, Jul 5, 2017 at 12:07 AM, Fernando Dodino <notifications@github.com>
wrote:

> Hi Pablo,
> I'm sorry but I can't figure out how to make it work. I ran Wollok
> Environment in Development Mode, and as I had some problems, I decided to
> delete the whole project and start a new fresh one. This is what I have
>
> [image: image]
> <https://user-images.githubusercontent.com/4549002/27843600-cd08a450-60eb-11e7-92c3-5150747a5462.png>
>
> I'm not sure what I'm doing wrong, but if I write "import " and then press
> Ctrl + Space I can see no packages (I added some packages from Wollok
> environment, but I think it is the same if you create a Folder in the IDE).
> If you want me to do further tests, I can try another situation.
>
> If I try to do
>
> import ../aves.aves.*
>
> I get a
>
> org.eclipse.emf.common.util.WrappedException: java.lang.ArrayIndexOutOfBoundsException
> 	at org.eclipse.xtext.linking.lazy.LazyLinkingResource.getEObject(LazyLinkingResource.java:230)
> 	at org.eclipse.xtext.linking.lazy.LazyLinkingResource.doResolveLazyCrossReference(LazyLinkingResource.java:189)
> 	at org.eclipse.xtext.linking.lazy.LazyLinkingResource.resolveLazyCrossReference(LazyLinkingResource.java:148)
> 	at org.eclipse.xtext.linking.lazy.LazyLinkingResource.resolveLazyCrossReferences(LazyLinkingResource.java:134)
> 	at org.eclipse.xtext.EcoreUtil2.resolveLazyCrossReferences(EcoreUtil2.java:498)
> 	at org.eclipse.xtext.resource.impl.DefaultResourceDescription.computeReferenceDescriptions(DefaultResourceDescription.java:135)
> 	at org.eclipse.xtext.resource.impl.DefaultResourceDescription$2.get(DefaultResourceDescription.java:123)
> 	at org.eclipse.xtext.resource.impl.DefaultResourceDescription$2.get(DefaultResourceDescription.java:120)
> 	at org.eclipse.xtext.util.OnChangeEvictingCache.get(OnChangeEvictingCache.java:77)
> 	at org.eclipse.xtext.resource.impl.DefaultResourceDescription.getReferenceDescriptions(DefaultResourceDescription.java:120)
> 	at org.eclipse.xtext.resource.DescriptionUtils.collectOutgoingReferences(DescriptionUtils.java:29)
> 	at org.eclipse.xtext.resource.impl.DefaultResourceDescriptionManager.isAffected(DefaultResourceDescriptionManager.java:117)
> 	at org.eclipse.xtext.ui.editor.DirtyStateEditorSupport.isReparseRequired(DirtyStateEditorSupport.java:649)
> 	at org.eclipse.xtext.ui.editor.DirtyStateEditorSupport$UpdateEditorStateJob$1.exec(DirtyStateEditorSupport.java:161)
> 	at org.eclipse.xtext.ui.editor.DirtyStateEditorSupport$UpdateEditorStateJob$1.exec(DirtyStateEditorSupport.java:1)
> 	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:91)
> 	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:520)
> 	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.readOnly(XtextDocument.java:492)
> 	at org.eclipse.xtext.ui.editor.model.XtextDocument.readOnly(XtextDocument.java:133)
> 	at org.eclipse.xtext.ui.editor.DirtyStateEditorSupport$UpdateEditorStateJob.run(DirtyStateEditorSupport.java:146)
> 	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
> Caused by: java.lang.ArrayIndexOutOfBoundsException
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/uqbar-project/wollok/pull/1086#issuecomment-312960845>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAVsf0qoZVXbKj9zSHxytuv7r4VmOR-4ks5sKre1gaJpZM4Na2LC>
> .
>



-- 
Pablo Tesone.
tesonep@gmail.com

tesonep commented:
I have fixed some problems when handling resources in the workspace. It was working only in the runtime. I have tested, but if anybody can check it again it will be nice.

![screen shot 2017-07-05 at 14 44 02](https://user-images.githubusercontent.com/355455/27864922-d3346fbc-6190-11e7-9019-97ed9f781342.png)

tesonep commented:
I still have problems with Travis now.

tesonep commented:
Ready. Now it is working... I think.

fdodino commented:
Great! I'm now working on another branch. As soon as I can will try to test it.

fdodino commented:
@tesonep , like a charm!

![image](https://user-images.githubusercontent.com/4549002/27912174-4a636790-6232-11e7-93fa-870c7ff65ff4.png)

Relative imports does not work, but I don't like them much. I prefer to use this way.
I did a quick test. Is there something more you want to test?

tesonep commented:
How are you testing the relative imports, maybe there is a difference in what is the expected syntax to write it. In the image I have captured, the file pruebas.wtest is importing aves.*, they are both in model. You are trying to use '..' or things like that? Is that what we want? Maybe we have different ideas.

fdodino commented:
Aaaaaaaaah, I misunderstood the concept of "relative". I was trying to use ../aves/... Obviously that's incorrect. Ok then.
